src/qibo/hamiltonians.py

```diff
@@ -186,3 +188,129 @@ def MaxCut(nqubits, dense=True):
+    print(distance_matrix)
+    small_tsp = TSP(distance_matrix)
+    obj_hamil, mixer = small_tsp.TspHamiltonians(dense=False)
+    initial_parameters = np.random.uniform(0, 1, 2)
+    initial_state = small_tsp.PrepareInitialStateTsp([i for i in range(num_cities)])
+    qaoa = QAOA(obj_hamil, mixer=mixer)
+
+    """
+    def __init__(self, distance_matrix):
+        """
+        Args:
+            distance_matrix: a numpy matrix encoding the distance matrix.
+        """
+        self.distance_matrix = distance_matrix
+        self.two_to_one = dict()
+        self.num_cities = distance_matrix.shape[0]
+        counter = 0
+        for i in range(self.num_cities):
+            for j in range(self.num_cities):
+                self.two_to_one[(i, j)] = counter
+                counter += 1
+
+
+    def PrepareObjTsp(self, dense=True):
+        """ This function returns the objective Hamiltonian
+        Args:
+            dense: Whether the Hamiltonian should be dense.
+
+        Returns: objective Hamiltonian that we wish to minimize the tsp distance.
+
+        """
+        form = 0
+        for i in range(self.num_cities):
+            for u in range(self.num_cities):
+                for v in range(self.num_cities):
+                    if u != v:
+                        form += self.distance_matrix[u, v] * tspZ(self.two_to_one[u, i]) * tspZ(self.two_to_one[v, (i + 1) % self.num_cities])
+        ham = SymbolicHamiltonian(form)
+        if dense:
+            ham = ham.dense
+        return ham
+
+
+    def SPlusTsp(self, u, i):
+        """ This is a subroutine required for the mixer performing X+iY
+
+        Args:
+            u: this represents the index of the object.
+            i: this represents the location of the slot
+
+        Returns: the gate for the subroutine
+        """
+        return tspX(self.two_to_one[u, i]) + 1j*tspY(self.two_to_one[u,i])
+
+
+    def SNegTsp(self, u, i):
+        """ This is a subroutine required for the mixer performing X-iY
+
+        Args:
+            u: this represents the index of the object.
+            i: this represents the location of the slot
+
+        Returns: the gate for the subroutine
+        """
+        return tspX(self.two_to_one[u, i]) - 1j*tspY(self.two_to_one[u,i])
+
+
+    def TspMixer(self, dense=True):
+        """
+
+        Args:
+            dense: Indicate whether the Hamiltonian is dense
+
+        Returns: Hamiltonian describing the mixer.
+
+        """
+        form = 0
+        for i in range(self.num_cities):
+            for u in range(self.num_cities):
+                for v in range(self.num_cities):
+                    if u != v:
+                        form += self.SPlusTsp(u, i) * self.SPlusTsp(v, (i+1) % self.num_cities) * self.SNegTsp(u, (i+1)%self.num_cities) * self.SNegTsp(v, i) + self.SNegTsp(u, i) * self.SNegTsp(v, (i+1) % self.num_cities) * self.SPlusTsp(u, (i+1) % self.num_cities) * self.SPlusTsp(v, i)
```

Comment:
Very long lines usually look better if split to multiple lines.

Reply:
Thank you so much for your suggestions and feedbac. Thanks for introducing me to Sphinx as well.  

 I have changed some files (and also introduce some bugs). Let me work on it a little bit more.

Reply:
Hello, I tried running QAOA and I obtained the following state vector output:

`(0.07208-0.29489j)|001010100> + (0.01032-0.13008j)|001100010> + (0.38305-0.14509j)|010001100> + (-0.37556-0.44498j)|010100001> + (0.60235+0.0927j)|100001010> + (-0.10676-0.03421j)|100010001>`

Any advice on  what I can do about it? such as how do I measure it multiple times? or convert it to a dictionary encoding the probability in Qibo?